docs/design/2022-03-09-optimize-gc-for-internal-transaction.md

```diff
@@ -0,0 +1,207 @@
+
+### data structure design
+
+#### SessionManager 
+
+TiDB manages client sessions by SessionManager interface. This design expands the SessionManager interface to manage internal sessions too.
+
+```go
+type SessionManager interface {
+	// Put the internal session pointer to the map in the SessionManager
+	StoreInternalSession(addr unsafe.Pointer)
+	// Delete the internal session pointer from the map in the SessionManager
+	DeleteInternalSession(addr unsafe.Pointer)
+	// Get all startTS of every transactions running in the current internal sessions
+	GetInternalSessionStartTSList() []uint64
+}
+```
+
+
+
+#### innerTxnStartTsBox 
+
+This design defines a global variable `globalInnerTxnTsBox` to manage internal transactions run by function `RunInNewTxn`. The map `innerTxnStartTsBox.innerTxnStartTsMap` stores all the internal transaction startTS. 
+
+```go
+var globalInnerTxnTsBox atomic.Value
+
+type innerTxnStartTsBox struct {
+	wg                  sync.WaitGroup
+	chanToStoreStartTS  chan uint64
+	chanToDeleteStartTS chan uint64
+
+	innerTxnTsBoxRun atomic.Value
+	
+	innerTSLock        sync.Mutex
+	innerTxnStartTsMap map[uint64]uint64
+	
+	// `storeInnerTxnStartTsLoop` and `deleteInnerTxnStartTsLoop` run asynchronously,It can't ensure  
+    //`storeInnerTxnStartTsLoop` receives startTS before `deleteInnerTxnStartTsLoop`,though 
+    // `RunInNewTxn`send startTS to `storeInnerTxnStartTsLoop` firstly. If `deleteInnerTxnStartTsLoop`
+	// recevied startTS before `storeInnerTxnStartTsLoop`, the `startTS` couldn't be deleted from 
+    // `innerTxnStartTsMap`,and GC safepoint can't be advanced properly.
+	undeletedTsLock        sync.Mutex
+	chanToProcUndelStartTS chan uint64
+	undeletedStartTsMap    map[uint64]uint64
+
+}
+```
+
+
+
+### Functional design
+
+#### Initialize internal transaction global management box
+
+This design initializes the global variable `globalInnerTxnTsBox` and starts 3 goroutines in the `main()` function to manage internal transactions run by function `RunInNewTxn()`. The initialization is wrapped in the function `InitInnerTxnStartTsBox()`. It closes the relevant goroutines in the function `cleanup()`
+
+```go
+func main() {
+	kv.InitInnerTxnStartTsBox()
+    signal.SetupSignalHandler(func(graceful bool) {
+        svr.Close()
+        cleanup(svr, storage, dom, graceful)
+        cpuprofile.StopCPUProfiler()
+        close(exited)
+    })
+}
+```
+
+
+
+the main code for `InitInnerTxnStartTsBox()` is as follower. It starts `storeInnerTxnStartTsLoop` goroutine to store an internal transaction startTS in `globalInnerTxnTsBox.innerTxnStartTsMap`, starts `deleteInnerTxnStartTsLoop` goroutine to delete an internal transaction startTS from `globalInnerTxnTsBox.innerTxnStartTsMap`. In particular, It starts `processUndeletedStartTsLoop` to delete the startTS for some rarely scenes.
```

Comment:
The two goroutines `storeInnerTxnStartTsLoop` and `deleteInnerTxnStartTsLoop` runs asynchronously. Though the `InnerTxnStartTs` is sent to `storeInnerTxnStartTsLoop` firstly and then sent to `deleteInnerTxnStartTsLoop` , But it can't ensure which goroutine received `InnerTxnStartTs`  firstly. If `deleteInnerTxnStartTsLoop` received `InnerTxnStartTs`  first, `InnerTxnStartTs`  cant be deleted forever